### PR TITLE
Implement CS checking based on the `WP_CLI_CS` ruleset

### LINKED ISSUE #104: Adopt and enforce new WPCliCS standard
We have a new PHPCS standard for WP-CLI called `WPCliCS` (props @jrfnl). It is part of the `wp-cli/wp-cli-tests` package starting with version v2.1.0.

To adopt & enforce this new standard, the following actions need to be taken for this repository:

- [ ] Create a PR that adds a custom ruleset `phpcs.xml.dist` to the repository

    - Add `phpcs.xml.dist` file
    - Adapt `.distignore` to ignore `phpcs.xml.dist` & `phpunit.xml.dist`
    - Adapt `.gitignore` to ignore `phpunit.xml`, `phpcs.xml` & `.phpcs.xml`
    - Require version `^2.1` of the `wp-cli/wp-cli-tests` as a dev dependency
- [ ] Make any required changes to the code that fail the checks from the above ruleset in separate PRs
- [ ] Merge thre ruleset once all required changes have been processed and merged

A sample PR for a simple repository can be seen here: https://github.com/wp-cli/maintenance-mode-command/pull/3/files

### PR DESCRIPTION
**Pull request by thrijith**
Add a PHPCS ruleset using the new `WPCliCS` standard.

Fixes #104

### CODE REVIEW

File: src/Media_Command.php [hunk 1]
```diff
@@ -117,7 +120,7 @@ function regenerate( $args, $assoc_args = array() ) {
 			}
 		}
 
-		$skip_delete = \WP_CLI\Utils\get_flag_value( $assoc_args, 'skip-delete' );
+		$skip_delete  = \WP_CLI\Utils\get_flag_value( $assoc_args, 'skip-delete' );
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
		$skip_delete  = Utils\get_flag_value( $assoc_args, 'skip-delete' );
```

File: src/Media_Command.php [hunk 2]
```diff
@@ -117,7 +120,7 @@ function regenerate( $args, $assoc_args = array() ) {
...
 		$only_missing = \WP_CLI\Utils\get_flag_value( $assoc_args, 'only-missing' );
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
		$only_missing = Utils\get_flag_value( $assoc_args, 'only-missing' );
```

File: src/Media_Command.php [hunk 3]
```diff
@@ -256,26 +270,28 @@ function import( $args, $assoc_args = array() ) {
...
+		$errors    = 0;
 		foreach ( $args as $file ) {
 			$number++;
 			if ( 0 === $number % self::WP_CLEAR_OBJECT_CACHE_INTERVAL ) {
 				Utils\wp_clear_object_cache();
 			}
-			$is_file_remote = parse_url( $file, PHP_URL_HOST );
-			$orig_filename = $file;
-			$file_time = '';
+			$is_file_remote = function_exists( 'wp_parse_url' ) ? wp_parse_url( $file, PHP_URL_HOST ) : parse_url( $file, PHP_URL_HOST ); // phpcs:ignore WordPress.WP.AlternativeFunctions.parse_url_parse_url -- parse_url will only be used in absence of wp_parse_url.
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
			// phpcs:ignore WordPress.WP.AlternativeFunctions.parse_url_parse_url -- parse_url will only be used in absence of wp_parse_url.
			$is_file_remote = function_exists( 'wp_parse_url' ) ? wp_parse_url( $file, PHP_URL_HOST ) : parse_url( $file, PHP_URL_HOST );
```

File: src/Media_Command.php [hunk 4]
```diff
@@ -384,17 +409,22 @@ function import( $args, $assoc_args = array() ) {
 			$attachment_success_text = '';
 			if ( $assoc_args['post_id'] ) {
 				$attachment_success_text = " and attached to post {$assoc_args['post_id']}";
-				if ( \WP_CLI\Utils\get_flag_value( $assoc_args, 'featured_image' ) )
+				if ( \WP_CLI\Utils\get_flag_value( $assoc_args, 'featured_image' ) ) {
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
				if ( Utils\get_flag_value( $assoc_args, 'featured_image' ) ) {
```

File: src/Media_Command.php [hunk 5]
```diff
@@ -384,17 +409,22 @@ function import( $args, $assoc_args = array() ) {
...
 					$attachment_success_text .= ' as featured image';
+				}
 			}
 
 			if ( \WP_CLI\Utils\get_flag_value( $assoc_args, 'porcelain' ) ) {
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
			if ( Utils\get_flag_value( $assoc_args, 'porcelain' ) ) {
```

File: src/Media_Command.php [hunk 6]
```diff
@@ -825,7 +873,7 @@ private function get_intermediate_sizes( $is_pdf, $metadata ) {
 
 		// Check here that not PDF (as filter not applied in core if is) and `$metadata` is array (as may not be and filter only applied in core when is).
 		if ( ! $is_pdf && is_array( $metadata ) ) {
-			$sizes = apply_filters( 'intermediate_image_sizes_advanced', $sizes, $metadata );
+			$sizes = apply_filters( 'intermediate_image_sizes_advanced', $sizes, $metadata ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound -- Calling native WordPress hook.
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
			// phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound -- Calling native WordPress hook.
			$sizes = apply_filters( 'intermediate_image_sizes_advanced', $sizes, $metadata );
```

File: src/Media_Command.php [hunk 7]
```diff
@@ -792,11 +841,10 @@ private function get_intermediate_sizes( $is_pdf, $metadata ) {
 			// For WP < 4.7.0.
 			global $_wp_additional_image_sizes;
 			if ( ! $_wp_additional_image_sizes ) {
-				$_wp_additional_image_sizes = array();
+				$_wp_additional_image_sizes = array(); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited -- Used as a fallback for WordPress version less than 4.7.0 as function wp_get_additional_image_sizes didn't exist then.
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited -- Used as a fallback for WordPress version less than 4.7.0 as function wp_get_additional_image_sizes didn't exist then.
				$_wp_additional_image_sizes = array();
```

File: src/Media_Command.php [hunk 8]
```diff
@@ -774,12 +823,12 @@ private function get_intermediate_image_sizes_for_attachment( $fullsizepath, $is
...
 		if ( $is_pdf ) {
 			// Copied from wp_generate_attachment_metadata() in "wp-admin/includes/image.php".
-			$fallback_sizes = array(
+			$fallback_sizes           = array(
 				'thumbnail',
 				'medium',
 				'large',
 			);
-			$intermediate_image_sizes = apply_filters( 'fallback_intermediate_image_sizes', $fallback_sizes, $metadata );
+			$intermediate_image_sizes = apply_filters( 'fallback_intermediate_image_sizes', $fallback_sizes, $metadata ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound -- Calling native WordPress hook.
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
			// phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound -- Calling native WordPress hook.
			$intermediate_image_sizes = apply_filters( 'fallback_intermediate_image_sizes', $fallback_sizes, $metadata );
```

File: src/Media_Command.php [hunk 9]
```diff
@@ -344,16 +363,19 @@ function import( $args, $assoc_args = array() ) {
 			}
 
 			if ( \WP_CLI\Utils\get_flag_value( $assoc_args, 'skip-copy' ) ) {
```
Comment (schlessera) [inline src/Media_Command.php]:
```suggestion
			if ( Utils\get_flag_value( $assoc_args, 'skip-copy' ) ) {
```

**Comment by schlessera** [comment]
Ok, the file is now being loaded twice, so you need to wrap a `function_exists()` check around the `wpcli_assert_image_editor_support()` function.

Silly, but there's no easy way I can think of to avoid that.

**Comment by schlessera** [comment]
Hmm, maybe it is worth investigating why it is being loaded twice after all...

**Comment by schlessera** [comment]
So wrapping the function in `function_exists()`, but it is just cloaking the fact that we're actually _adding_  the command twice:

Image 2019-04-19 at 7 40 18 PM

This is of course just the case when doing development on the command itself, but nevertheless, this bothers me.

Two solutions I can think of:
- Define a new pattern to make sure the command bootstrap file is only processed once (will probably need to be `define` based).
- Bail early if you try to `add_command()` a command that was already added (will need to investigate what this means for overrides).

**Comment by schlessera** [comment]
Here's what I came up with so far.

This solves two problems:
- the file can be _parsed_ multiple times without throwing an error (because I turned the regular function into a closure)
- the file can be _executed_ multiple times without adding the command more than once (because of the define).

```php
<?php

if ( defined( 'WP_CLI_MEDIA_COMMAND_LOADED' ) || ! class_exists( 'WP_CLI' ) ) {
	return;
}

define( 'WP_CLI_MEDIA_COMMAND_LOADED', true );

$wpcli_media_autoloader = dirname( __FILE__ ) . '/vendor/autoload.php';
if ( file_exists( $wpcli_media_autoloader ) ) {
	require_once $wpcli_media_autoloader;
}

/**
 * Check for Imagick and GD extensions availability.
 *
 * @throws \WP_CLI\ExitException
 */
$wpcli_media_assert_image_editor_support = function () {
	if ( ! wp_image_editor_supports() ) {
		WP_CLI::error(
			'No support for generating images found. '
			. 'Please install the Imagick or GD PHP extensions.'
		);
	}
};

WP_CLI::add_command(
	'media',
	'Media_Command',
	[ 'before_invoke' => $wpcli_media_assert_image_editor_support ]
);
```

How do you feel about that?

**Comment by schlessera** [comment]
I can't really make up my mind just yet on a new pattern, so let's ignore the double-loading for this release.

To fix the broken tests, just turn the regular function into a closure:
```php
/**
 * Check for Imagick and GD extensions availability.
 *
 * @throws \WP_CLI\ExitException
 */
$wpcli_media_assert_image_editor_support = function () {
	if ( ! wp_image_editor_supports() ) {
		WP_CLI::error(
			'No support for generating images found. '
			. 'Please install the Imagick or GD PHP extensions.'
		);
	}
};

WP_CLI::add_command(
	'media',
	'Media_Command',
	[ 'before_invoke' => $wpcli_media_assert_image_editor_support ]
);
```

**Comment by thrijith** [comment]
I was away so couldn't reply faster, this seems fine https://github.com/wp-cli/media-command/pull/106#issuecomment-484970453.

> so let's ignore the double-loading for this release.

Didn't get this part.

So, let's go with the last option then? The invoking function was a closure already, we now have better readability.

**Pull request merged**